File: backend/app/schemas/hermes.py
"""Contracts for the project-scoped Hermes retrieval assistant."""

import json
from datetime import date, datetime
from typing import Literal
from uuid import uuid4

from pydantic import BaseModel, Field, field_validator, model_validator

from app.schemas.hermes_tools import HermesToolName


HermesSourceType = Literal["knowledge", "requirement", "case"]
HermesMessageRole = Literal["user", "assistant"]


class HermesEvaluationScoresOut(BaseModel):
    tool_selection: bool | None = None
    citation_relevance: bool | None = None
    answer_completeness: bool | None = None
    refusal_correctness: bool


class HermesEvaluationResultOut(BaseModel):
    set_id: str
    set_version: str
    case_id: str
    scores: HermesEvaluationScoresOut


class HermesHistoryItem(BaseModel):
    role: HermesMessageRole
    content: str = Field(min_length=1, max_length=2_000)

    @field_validator("content")
    @classmethod
    def trim_content(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("对话内容不能为空")
        return value


class HermesQueryIn(BaseModel):
    project_id: int = Field(ge=1)
    query: str = Field(min_length=1, max_length=2_000)
    limit: int = Field(default=8, ge=1, le=20)
    conversation_id: str = Field(
        default_factory=lambda: f"hermes-{uuid4().hex}",
        min_length=8,
        max_length=80,
        pattern=r"^[A-Za-z0-9._:-]+$",
    )
    history: list[HermesHistoryItem] = Field(default_factory=list, max_length=12)
    source_types: list[HermesSourceType] = Field(default_factory=list, max_length=3)
    updated_from: date | None = None
    updated_to: date | None = None
    context_budget: int = Field(default=6_000, ge=1_000, le=12_000)
    session_id: int | None = Field(default=None, ge=1)

    @field_validator("query")
    @classmethod
    def trim_query(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("查询内容不能为空")
        return value

    @field_validator("conversation_id", mode="before")
    @classmethod
    def trim_conversation_id(cls, value: object) -> object:
        return value.strip() if isinstance(value, str) else value

    @field_validator("source_types")
    @classmethod
    def normalize_source_types(cls, values: list[HermesSourceType]) -> list[HermesSourceType]:
        return list(dict.fromkeys(values))

    @model_validator(mode="after")
    def validate_date_range(self) -> "HermesQueryIn":
        if self.updated_from and self.updated_to and self.updated_from > self.updated_to:
            raise ValueError("更新时间范围无效")
        return self


class HermesSourceOut(BaseModel):
    source_type: HermesSourceType
    source_id: int
    project_id: int | None = None
    title: str
    excerpt: str
    source_ref: str | None = None
    path: str
    match_terms: list[str] = Field(default_factory=list)
    match_score: int = Field(ge=0)
    updated_at: datetime | None = None


class HermesQueryOut(BaseModel):
    project_id: int
    query: str
    conversation_id: str
    history_used: int = Field(ge=0)
    history_omitted: int = Field(ge=0)
    context_chars: int = Field(ge=0)
    context_budget: int = Field(ge=1_000, le=12_000)
    source_types: list[HermesSourceType] = Field(default_factory=list, max_length=3)
    updated_from: date | None = None
    updated_to: date | None = None
    mode: Literal["llm_grounded", "project_retrieval", "no_results"]
    answer: str
    sources: list[HermesSourceOut] = Field(default_factory=list, max_length=20)
    generated_at: datetime
    session_id: int
    message_index: int
    prompt_version: str = "hermes-v2"
    latency_ms: int = Field(ge=0)
    evaluation: HermesEvaluationResultOut | None = None


class HermesSessionOut(BaseModel):
    id: int
    project_id: int
    title: str
    context_filters: dict
    messages: list[dict]
    drafts: list[dict]
    metrics: dict
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class HermesSessionCreateIn(BaseModel):
    project_id: int = Field(ge=1)
    title: str = Field(default="Hermes 会话", max_length=80)

    @field_validator("title")
    @classmethod
    def trim_title(cls, value: str) -> str:
        return value.strip() or "Hermes 会话"


class HermesEvaluationQuestionOut(BaseModel):
    id: str
    prompt: str
    expected_mode: Literal["llm_grounded", "project_retrieval", "no_results"]
    execution: Literal["query", "orchestrate"]
    expected_tools: list[HermesToolName] = Field(default_factory=list, max_length=2)
    expected_source_types: list[HermesSourceType] = Field(default_factory=list, max_length=3)
    requires_citation: bool
    required_answer_terms: list[str] = Field(default_factory=list, max_length=5)
    expected_refusal: bool


class HermesEvaluationSetOut(BaseModel):
    id: str
    version: str
    questions: list[HermesEvaluationQuestionOut] = Field(default_factory=list, max_length=20)


class HermesEvaluationSetMetaOut(BaseModel):
    id: str
    version: str
    size: int = Field(ge=0)


class HermesCostTrackingOut(BaseModel):
    available: bool
    reason: (
        Literal[
            "no_model_calls",
            "usage_unavailable",
            "pricing_not_configured",
            "partially_unpriced",
        ]
        | None
    ) = None
    amounts_by_currency: dict[str, float] = Field(default_factory=dict)
    priced_calls: int = Field(default=0, ge=0)
    unpriced_calls: int = Field(default=0, ge=0)


class HermesModelPlanningGovernanceOut(BaseModel):
    attempts: int = Field(ge=0)
    model_calls: int = Field(ge=0)
    usage_calls: int = Field(ge=0)
    input_tokens: int = Field(ge=0)
    output_tokens: int = Field(ge=0)
    total_tokens: int = Field(ge=0)
    average_latency_ms: int = Field(ge=0)
    fallback_count: int = Field(ge=0)
    fallback_reasons: dict[str, int] = Field(default_factory=dict)


class HermesEvaluationMetricOut(BaseModel):
    evaluated: int = Field(ge=0)
    passed: int = Field(ge=0)
    rate: float | None = Field(default=None, ge=0, le=1)


class HermesEvaluationQualityOut(BaseModel):
    runs: int = Field(ge=0)
    cases_covered: int = Field(ge=0)
    tool_selection: HermesEvaluationMetricOut
    citation_relevance: HermesEvaluationMetricOut
    answer_completeness: HermesEvaluationMetricOut
    refusal_correctness: HermesEvaluationMetricOut


class HermesGovernanceSummaryOut(BaseModel):
    prompt_version: str
    prompt_versions: list[str] = Field(default_factory=list, max_length=20)
    evaluation_set: HermesEvaluationSetMetaOut
    sessions: int = Field(ge=0)
    assistant_messages: int = Field(ge=0)
    citation_coverage: float = Field(ge=0, le=1)
    refusal_rate: float = Field(ge=0, le=1)
    no_result_rate: float = Field(ge=0, le=1)
    helpful_count: int = Field(ge=0)
    not_helpful_count: int = Field(ge=0)
    feedback_total: int = Field(ge=0)
    helpful_rate: float | None = Field(default=None, ge=0, le=1)
    average_latency_ms: int = Field(ge=0)
    p95_latency_ms: int = Field(ge=0)
    evaluation_quality: HermesEvaluationQualityOut
    model_planning: HermesModelPlanningGovernanceOut
    cost_tracking: HermesCostTrackingOut


class HermesToolIn(BaseModel):
    project_id: int = Field(ge=1)
    arguments: dict = Field(default_factory=dict)


class HermesDraftIn(BaseModel):
    project_id: int = Field(ge=1)
    draft_type: Literal["test_plan"] = "test_plan"
    payload: dict
    sources: list[dict] = Field(default_factory=list, max_length=20)

    @field_validator("payload")
    @classmethod
    def bound_payload(cls, value: dict) -> dict:
        if len(json.dumps(value, ensure_ascii=False, default=str).encode("utf-8")) > 64 * 1024:
            raise ValueError("Hermes 草稿不能超过 64KB")
        return value


class HermesDraftConfirmIn(BaseModel):
    project_id: int = Field(ge=1)
    draft_id: str = Field(min_length=1, max_length=64)
    confirmation: Literal["CONFIRM"]


class HermesFeedbackIn(BaseModel):
    project_id: int = Field(ge=1)
    message_index: int = Field(ge=0)
    rating: Literal["helpful", "not_helpful"]
    comment: str | None = Field(default=None, max_length=1000)

File: frontend/src/api/hermes.ts
import http from './http'

export type HermesSourceType = 'knowledge' | 'requirement' | 'case'

export interface HermesEvaluationResult {
  set_id: string
  set_version: string
  case_id: string
  scores: {
    tool_selection: boolean | null
    citation_relevance: boolean | null
    answer_completeness: boolean | null
    refusal_correctness: boolean
  }
}

export interface HermesSourceItem {
  source_type: HermesSourceType
  source_id: number
  project_id?: number | null
  title: string
  excerpt: string
  source_ref?: string | null
  path: string
  match_terms: string[]
  match_score: number
  updated_at?: string | null
}

export interface HermesQueryResult {
  project_id: number
  query: string
  conversation_id: string
  history_used: number
  history_omitted: number
  context_chars: number
  context_budget: number
  source_types: HermesSourceType[]
  updated_from?: string | null
  updated_to?: string | null
  mode: 'llm_grounded' | 'project_retrieval' | 'no_results'
  answer: string
  sources: HermesSourceItem[]
  generated_at: string
  session_id: number
  message_index: number
  prompt_version: string
  latency_ms: number
  evaluation?: HermesEvaluationResult | null
}

export interface HermesSessionItem {
  id: number
  project_id: number
  title: string
  context_filters: Record<string, unknown>
  messages: Array<Record<string, unknown>>
  drafts: Array<Record<string, unknown>>
  metrics: Record<string, unknown>
  state_version: number
  created_at: string
  updated_at: string
}

export interface HermesGovernanceSummary {
  prompt_version: string
  prompt_versions: string[]
  evaluation_set: { id: string; version: string; size: number }
  sessions: number
  assistant_messages: number
  citation_coverage: number
  refusal_rate: number
  no_result_rate: number
  helpful_count: number
  not_helpful_count: number
  feedback_total: number
  helpful_rate: number | null
  average_latency_ms: number
  p95_latency_ms: number
  evaluation_quality: {
    runs: number
    cases_covered: number
    tool_selection: HermesEvaluationMetric
    citation_relevance: HermesEvaluationMetric
    answer_completeness: HermesEvaluationMetric
    refusal_correctness: HermesEvaluationMetric
  }
  model_planning: {
    attempts: number
    model_calls: number
    usage_calls: number
    input_tokens: number
    output_tokens: number
    total_tokens: number
    average_latency_ms: number
    fallback_count: number
    fallback_reasons: Record<string, number>
  }
  cost_tracking: {
    available: boolean
    reason?: 'no_model_calls' | 'usage_unavailable' | 'pricing_not_configured' | 'partially_unpriced' | null
    amounts_by_currency: Record<string, number>
    priced_calls: number
    unpriced_calls: number
  }
}

export interface HermesEvaluationMetric {
  evaluated: number
  passed: number
  rate: number | null
}

export type HermesToolName =
  | 'failed_tasks'
  | 'run_detail'
  | 'quality_trend'
  | 'requirement_case_links'
  | 'knowledge_detail'
export type HermesToolStatus = 'ok' | 'empty' | 'not_found' | 'timeout' | 'error'

export interface HermesToolDescriptor {
  name: HermesToolName
  description: string
  required_role: 'viewer'
  read_only: true
  timeout_max_ms: number
  arguments_schema: Record<string, unknown>
}

export interface HermesToolEvidence {
  evidence_id: string
  source_type: 'hermes_tool'
  source_ref: string
  title: string
  excerpt: string
  path: string
}

export interface HermesToolResult {
  project_id: number
  conversation_id: string
  tool: HermesToolName
  status: HermesToolStatus
  duration_ms: number
  message?: string | null
  data: Record<string, unknown>
  evidence: HermesToolEvidence[]
  generated_at: string
}

export interface HermesToolCall {
  project_id: number
  conversation_id: string
  tool: HermesToolName
  arguments?: Record<string, unknown>
  timeout_ms?: number
}

export interface HermesOrchestrationPlan {
  tool: HermesToolName
  arguments: Record<string, unknown>
  reason: string
}

export interface HermesOrchestrationStep {
  tool: HermesToolName
  arguments: Record<string, unknown>
  status: HermesToolStatus
  duration_ms: number
  message?: string | null
  data: Record<string, unknown>
  evidence: HermesToolEvidence[]
}

export interface HermesPlannerDecision {
  source: 'deterministic' | 'model' | 'deterministic_fallback'
  validation: 'not_attempted' | 'accepted' | 'unavailable' | 'rejected' | 'failed'
  model_name?: string | null
  prompt_version?: string | null
  fallback_reason?: string | null
  model_calls: number
  latency_ms: number
  usage_available: boolean
  input_tokens?: number | null
  output_tokens?: number | null
  total_tokens?: number | null
  estimated_cost?: number | null
  currency?: string | null
}

export interface HermesOrchestrationResult {
  project_id: number
  conversation_id: string
  query: string
  status: 'matched' | 'no_match' | 'needs_input' | 'cancelled'
  clarification?: string | null
  plans: HermesOrchestrationPlan[]
  steps: HermesOrchestrationStep[]
  planner: HermesPlannerDecision
  answer: string
  generated_at: string
  session_id?: number | null
  message_index?: number | null
  evaluation?: HermesEvaluationResult | null
}

export interface HermesQueryRequest {
  project_id: number
  query: string
  limit?: number
  conversation_id?: string
  history?: Array<{ role: 'user' | 'assistant'; content: string }>
  source_types?: HermesSourceType[]
  updated_from?: string
  updated_to?: string
  context_budget?: number
  session_id?: number
}

export interface HermesOrchestrationRequest {
  project_id: number
  query: string
  conversation_id?: string
  session_id?: number
}

export interface HermesDraftRequest {
  project_id: number
  draft_type: 'test_plan'
  payload: Record<string, unknown>
  sources?: Array<Record<string, unknown>>
}

export interface HermesFeedbackRequest {
  project_id: number
  message_index: number
  rating: 'helpful' | 'not_helpful'
  comment?: string
}

export const hermesApi = {
  query: (body: HermesQueryRequest) => http.post<unknown, HermesQueryResult>('/hermes/query', body),
  listTools: () => http.get<unknown, { tools: HermesToolDescriptor[]; generated_at: string }>('/hermes/tools'),
  executeTool: (body: HermesToolCall) => http.post<unknown, HermesToolResult>('/hermes/tools/execute', body),
  orchestrate: (body: HermesOrchestrationRequest) =>
    http.post<unknown, HermesOrchestrationResult>('/hermes/orchestrate', body),
  sessions: (projectId: number) =>
    http.get<unknown, HermesSessionItem[]>('/hermes/sessions', { params: { project_id: projectId } }),
  createSession: (projectId: number, title: string) =>
    http.post<unknown, HermesSessionItem>('/hermes/sessions', { project_id: projectId, title }),
  createDraft: (sessionId: number, body: HermesDraftRequest) =>
    http.post<unknown, { id: string; status: string }>(`/hermes/sessions/${sessionId}/drafts`, body),
  confirmDraft: (sessionId: number, body: { project_id: number; draft_id: string; confirmation: 'CONFIRM' }) =>
    http.post<unknown, { draft_id: string; status: string; plan_id: number }>(
      `/hermes/sessions/${sessionId}/drafts/confirm`,
      body,
    ),
  feedback: (sessionId: number, body: HermesFeedbackRequest) =>
    http.post(`/hermes/sessions/${sessionId}/feedback`, body),
  tool: (
    sessionId: number,
    toolName: 'failed_runs' | 'quality_summary',
    body: { project_id: number; arguments?: Record<string, unknown> },
  ) => http.post(`/hermes/sessions/${sessionId}/tools/${toolName}`, body),
  governance: (projectId: number) =>
    http.get<unknown, HermesGovernanceSummary>('/hermes/governance/summary', { params: { project_id: projectId } }),
}
